src/server.rs

```diff
@@ -336,49 +341,57 @@ async fn list(config: web::Data<RwLock<Config>>) -> Result<HttpResponse, Error>
-        .filter_map(|entry| {
-            entry.ok().and_then(|e| {
-                let metadata = match e.metadata() {
-                    Ok(metadata) => {
-                        if metadata.is_dir() {
-                            return None;
-                        }
-                        metadata
-                    }
-                    Err(e) => {
-                        error!("failed to read metadata: {e}");
-                        return None;
-                    }
-                };
-                let mut file_name = PathBuf::from(e.file_name());
-                let expires_at_utc = if let Some(expiration) = file_name
-                    .extension()
-                    .and_then(|ext| ext.to_str())
-                    .and_then(|v| v.parse::<i64>().ok())
-                {
-                    file_name.set_extension("");
-                    if util::get_system_time().ok()?
-                        > Duration::from_millis(expiration.try_into().ok()?)
-                    {
-                        return None;
-                    }
-                    Some(uts2ts::uts2ts(expiration / 1000).as_string())
-                } else {
-                    None
-                };
-                Some(ListItem {
-                    file_name,
-                    file_size: metadata.len(),
-                    expires_at_utc,
-                })
-            })
-        })
-        .collect();
+
+    let mut entries: Vec<ListItem> = Vec::new();
+
+    let mut dir_contents = fs::read_dir(config.server.upload_path).await?;
+
+    let system_time = util::get_system_time()?;
+
+    while let Ok(Some(entry)) = dir_contents.next_entry().await {
+        let metadata = match entry.metadata().await {
+            Ok(metadata) if metadata.is_dir() => continue,
+            Ok(metadata) => metadata,
+            Err(e) => {
+                error!("failed to read metadata: {e}");
+                continue;
+            }
+        };
+
+        let mut file_name = PathBuf::from(entry.file_name());
+
+        let expires_at_utc = match file_name
+            .extension()
+            .and_then(|ext| ext.to_str())
+            .and_then(|v| v.parse::<u64>().ok())
+        {
+            Some(expiration) if system_time > Duration::from_millis(expiration) => continue,
+            Some(expiration) => {
+                file_name.set_extension("");
+                Some(
+                    uts2ts::uts2ts(
+                        (expiration / 1000)
+                            .try_into()
+                            .map_err(|_| error::ErrorInternalServerError("Invalid timestamp"))?,
+                    )
+                    .as_string(),
+                )
+            }
+            None => None,
+        };
+
+        entries.push(ListItem {
+            file_name,
+            file_size: metadata.len(),
+            expires_at_utc,
+        });
+    }
+
```

Comment:
Is this refactored for async code? If so, we can probably extract this into a function.

Reply:
It's mostly refactored because the async versions of iterators aren't quite as powerful / there yet verses the sync counterparts. Using a loop like this is the standard way of consuming a `Stream`